### PR TITLE
Feature/tasking

### PR DESCRIPTION
**Pull request by bmcdanie**
break problems across both dimensions (elements and connections) to provide fine-grained discretization. 

specify a task size and the problem will be broken into subproblems requiring between 50% and 100% of the provided size (in memory). will enable upcoming gpu/mpi work.

### CODE REVIEW

**Comment by grahamlopez** [comment]
okay to test

File: src/grouping_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,805 @@
+
+#include "grouping.hpp"
+#include "tests_general.hpp"
+
+// check for complete, non-overlapping element assignment
+auto const validity_check = [](std::vector<element_group> const &groups,
```
Comment (grahamlopez) [inline src/grouping_tests.cpp]: why is this a lambda?

File: src/grouping_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,805 @@
...
+  {
+    for (int j = 0; j < coverage.ncols(); ++j)
+    {
+      REQUIRE(coverage(i, j) == element_status::assigned);
+    }
+  }
+};
+
+// check that a given task vector occupies between 49% and 101% of the limiit
+auto const size_check = [](std::vector<element_group> const &groups,
```
Comment (grahamlopez) [inline src/grouping_tests.cpp]: why is this a lambda?

File: src/grouping_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,805 @@
...
+      assert(groups.size() % ranks == 0);
+      assert(static_cast<int>(groups.size()) == num_groups);
+      validity_check(groups, table);
+      size_check(groups, *pde, limit_MB, large_problem);
+    }
+  }
+}
+
+// FIXME we should eventually put this in the pde class
+auto const element_segment_size = [](auto const &pde) {
```
Comment (grahamlopez) [inline src/grouping_tests.cpp]: this is repeated from `grouping.cpp:49`

File: src/grouping_tests.cpp (hunk 2 above)
Comment (grahamlopez) [reply src/grouping_tests.cpp]: and others as well

File: src/time_advance.cpp [hunk 4]
```diff
@@ -4,32 +4,18 @@
...
-
-  for (auto const &ops_list : batches)
-  {
-    assert(static_cast<int>(ops_list.size()) == pde.num_dims);
-    for (batch_operands_set<P> ops : ops_list)
-    {
-      assert(ops.size() == 3);
-    }
-  }
+  assert(dt >= 0);
```
Comment (grahamlopez) [inline src/time_advance.cpp]: @dlg0 are we allowing breaking of time-reversal symmetry? :-D
or is this merely for thermodynamic reasons :-(

File: src/time_advance.hpp [hunk 5]
```diff
@@ -1,26 +1,31 @@
...
 #include "batch.hpp"
+#include "grouping.hpp"
 #include "program_options.hpp"
 #include "tensors.hpp"
 
 // this function executes a time step using the current solution
 // vector x. on exit, the next solution vector is stored in fx.
 template<typename P>
-void explicit_time_advance(PDE<P> const &pde,
+void explicit_time_advance(PDE<P> const &pde, element_table const &table,
```
Comment (grahamlopez) [inline src/time_advance.hpp]: should there be an `element_table.hpp` include?

Is there any way to avoid adding this dependency?

File: src/time_advance.cpp [hunk 6]
```diff
@@ -41,40 +27,41 @@ void explicit_time_advance(PDE<P> const &pde,
   P const c2  = 1.0 / 2.0;
   P const c3  = 1.0;
 
-  apply_explicit(batches, system);
-  scale_sources(pde, unscaled_sources, system.scaled_source, time);
-  fm::axpy(system.scaled_source, system.batch_output);
-  fm::copy(system.batch_output, system.result_1);
+  apply_explicit(pde, table, groups, host_space, rank_space);
```
Comment (grahamlopez) [inline src/time_advance.cpp]: should there be an `element_table.hpp` include?

Is there any way to avoid adding this component dependency?

File: CMakeLists.txt [hunk 7]
```diff
@@ -64,13 +64,14 @@ set (components
   coefficients
   connectivity
   element_table
-  fast_math
+  fast_math 
+  grouping
```
[the remark this answers is not in the document]
Comment (grahamlopez) [reply CMakeLists.txt]: also, "group" is fine. just curious about the change of heart.

[the remark this answers is not in the document]
Comment (bmcdanie) [reply CMakeLists.txt]: I think I'd changed from tasking to grouping before the big "chunk" craze...

File: src/fast_math.hpp [hunk 8]
```diff
@@ -28,7 +28,7 @@ axpy(fk::vector<P, omem> const &x, fk::vector<P, mem> &y, P const alpha = 1.0)
 template<typename P, mem_type mem, mem_type omem>
 fk::vector<P, mem> &copy(fk::vector<P, omem> const &x, fk::vector<P, mem> &y)
 {
-  assert(x.size() == y.size());
+  assert(y.size() >= x.size());
```
[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/fast_math.hpp]: yes - blas only takes pointers x and y, their increments between elements, and the size of the copy. so, I don't think it is semantically incorrect to copy a smaller vector into a segment of a larger one.

File: src/grouping.cpp [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+  // calc in and out vector sizes for each elem
+  // since we will scheme to have most elems require overlapping pieces of
+  // x and y, we will never need 2 addtl xy space per elem
+  double const elem_xy_space_MB = get_MB(elem_size * 1.2);
+  return (elem_reduction_space_MB + elem_intermediate_space_MB +
+          elem_xy_space_MB);
+}
+
+// determine how many groups will be required to solve the problem
+// a task is a subset of all elements whose total workspace requirement
```
[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/grouping.cpp]: yes, fixed

File: src/grouping.hpp [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    int64_t num_elems = batch_input.size() + reduction_space.size() +
+                        batch_intermediate.size() + batch_output.size() +
+                        unit_vector_.size();
+    double const bytes     = static_cast<double>(num_elems) * sizeof(P);
+    double const megabytes = bytes * 1e-6;
+    return megabytes;
+  };
+
+private:
+  fk::vector<P> unit_vector_;
```
[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/grouping.hpp]: it isn't changed throughout the simulation - it should always be all ones. wanted to ensure it isn't mistakenly mis-set to something else. could do const, except that I don't have the size computed when it's initially constructed.

File: CMakeLists.txt [hunk 11]
```diff
@@ -64,13 +64,14 @@ set (components
...
   lib_dispatch
   matlab_utilities
   pde
   permutations
   program_options
-  quadrature
+  quadrature
```
[the remark this answers is not in the document]
Comment (bmcdanie) [reply CMakeLists.txt]: fixed

File: src/grouping.cpp [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+    int const groups_per_rank =
+        static_cast<int>(problem_size_per_rank / num_ranks + 1);
+    return groups_per_rank * num_ranks;
+  }();
+
+  return num_groups;
+}
+
+// divide the problem given the previously computed number of groups
+// this function divides via a greedy, row-major split.
```
[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/grouping.cpp]: yes - comments added

File: src/main.cpp [hunk 13]
```diff
@@ -147,46 +150,40 @@ int main(int argc, char **argv)
...
+            << get_MB(rank_space.reduction_space.size()) << '\n';
   std::cout << "kronmult working space size (MB): "
-            << get_MB(system.batch_intermediate.size()) << '\n';
-  std::cout << "output vector size (MB): " << get_MB(system.batch_output.size())
-            << '\n';
-  auto const &unit_vect = system.get_unit_vector();
+            << get_MB(rank_space.batch_intermediate.size()) << '\n';
+  std::cout << "output vector size (MB): "
+            << get_MB(rank_space.batch_output.size()) << '\n';
+  auto const &unit_vect = rank_space.get_unit_vector();
```
[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/main.cpp]: I felt rank space was more descriptive than explicit system - open to other views, though

File: src/grouping.cpp [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+          .second.second;
+  return std::make_pair(min_col, max_col);
+}
+std::pair<int, int> rows_in_group(element_group const &g)
+{
+  assert(g.size() > 0);
+  return std::make_pair(g.begin()->first, g.rbegin()->first);
+}
+
+auto const element_segment_size = [](auto const &pde) {
```
[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/grouping.cpp]: it does not have to be - it's just a small snippet that is used multiple places here and elsewhere. it may be helpful to have this as a pde class member.

File: src/grouping_tests.cpp (hunk 3 above)
Comment (bmcdanie) [reply src/grouping_tests.cpp]: yes - that lambda isn't available in this compilation unit. maybe it should be a function?

File: src/grouping_tests.cpp (hunk 2 above)
Comment (bmcdanie) [reply src/grouping_tests.cpp]: could be a function - just a test I want to apply multiple times. could write it as a function - I think the catch macros make it weird to define them in line here.

File: src/grouping_tests.cpp (hunk 1 above)
Comment (bmcdanie) [reply src/grouping_tests.cpp]: reused snippet within this scope, and catch makes defining functions inline here weird

File: src/grouping_tests.cpp (hunk 3 above)
Comment (bmcdanie) [reply src/grouping_tests.cpp]: defined in header

File: src/time_advance.cpp (hunk 4 above)
Comment (bmcdanie) [reply src/time_advance.cpp]: fixed

File: src/time_advance.hpp (hunk 5 above)
Comment (bmcdanie) [reply src/time_advance.hpp]: introduced by building batches here, and unfortunately, can't think of a way to get around it. header added.

File: src/time_advance.cpp (hunk 6 above)
Comment (bmcdanie) [reply src/time_advance.cpp]: dup?

File: CMakeLists.txt [hunk 15]
```diff
@@ -93,14 +94,16 @@ endif ()
 
 target_link_libraries (basis PRIVATE matlab_utilities quadrature tensors)
 
-target_link_libraries (batch PRIVATE lib_dispatch coefficients connectivity element_table pde tensors)
+target_link_libraries (batch PRIVATE lib_dispatch coefficients connectivity grouping element_table pde tensors)
```
Comment (grahamlopez) [inline CMakeLists.txt]: I think we want `INTERFACE chunk` for `time_advance` as well

File: src/grouping.hpp [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+#include "pde.hpp"
+#include "tensors.hpp"
+
+using element_group = std::map<int, std::pair<int, int>>;
+
+// convenience functions when working with element groups
+int num_elements_in_group(element_group const &g);
+int max_connected_in_group(element_group const &g);
+
+std::pair<int, int> columns_in_group(element_group const &g);
```
[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/grouping.hpp]: yes a small two-member struct may be more appropriate here, or rename. the pair is a start and stop index in both cases.

[the remark this answers is not in the document]
Comment (bmcdanie) [reply src/grouping.hpp]: I'd suggest `range` struct?

File: CMakeLists.txt [hunk 17]
```diff
@@ -93,14 +94,16 @@ endif ()
...
-target_link_libraries (batch PRIVATE lib_dispatch coefficients connectivity element_table pde tensors)
+target_link_libraries (batch PRIVATE lib_dispatch coefficients connectivity chunk element_table pde tensors)
 
 target_link_libraries (coefficients
   PRIVATE pde matlab_utilities quadrature tensors transformations)
 
 target_link_libraries (connectivity
   PRIVATE element_table matlab_utilities permutations tensors)
 
+target_link_libraries(chunk PRIVATE element_table fast_math pde)
```
Comment (grahamlopez) [inline CMakeLists.txt]: I think we need `INTERFACE element_table` on `time_advance` as well

File: src/element_table.hpp [hunk 18]
```diff
@@ -16,7 +16,7 @@
 // - coordinates are composed of a set of dimension-many pairs (level, cell).
 //   (technically, an element's coordinates will also have a degree component,
 //   but we omit this as we assume uniform degree).
-// - in our table, coordinates are stored with all level components grouped,
+// - in our table, coordinates are stored with all level components chunked,
```
Comment (grahamlopez) [inline src/element_table.hpp]: probably revert this change?

File: src/io.hpp [hunk 19]
```diff
@@ -25,7 +25,7 @@ HighFive::DataSet initialize_output_file(fk::vector<P> const &vec)
   HighFive::DataSpace dataspace = HighFive::DataSpace(
       {1, vec_size}, {HighFive::DataSpace::UNLIMITED, vec_size});
 
-  // Use chunking
+  // Use chunk
```
Comment (grahamlopez) [inline src/io.hpp]: probably revert this change?

File: src/main.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (grahamlopez) [reply src/main.cpp]: Sorry, I meant `unit_vector` -> `unit_vect`

File: src/matlab_utilities_tests.cpp [hunk 20]
```diff
@@ -190,7 +190,7 @@ TEMPLATE_TEST_CASE("find function", "[matlab]", float, double, int)
     fk::vector<int> gold;
     REQUIRE(find(haystack, greater_eq) == gold);
   }
-  SECTION("find a group -- vector")
+  SECTION("find a chunk -- vector")
```
Comment (grahamlopez) [inline src/matlab_utilities_tests.cpp]: revert this change?

Comment (grahamlopez) [reply src/matlab_utilities_tests.cpp]: (2 in this file)

File: src/time_advance.cpp (hunk 4 above)
Comment (grahamlopez) [reply src/time_advance.cpp]: @dlg0 says no reason for this other than "sanity"
